File: Threadily.test.web/AppExampleUnitTest.js
var assert = require('assert');
var module = require('./Threadily.SampleObjects');

describe('App Example', function () {

    it('Factory creates App, App creates Business, Business reads products', function (done) {
        var handle = null;

        var createProduct = function (newBusiness, productName, onDone) {
            handle = newBusiness.isCreateProductPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        newBusiness.isCreateProductPending.unsubscribe(handle);
                        console.log("Reading products complete");
                        onDone();
                    }
                }
            }));
            console.log("Creating product");
            newBusiness.createProductAsync(productName);
        }
        var readProducts = function (newBusiness, queryString, onDone) {
            handle = newBusiness.isProductsPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        newBusiness.isProductsPending.unsubscribe(handle);
                        console.log("Reading products complete");
                        onDone();
                    }
                }
            }));
            console.log("Reading products");
            newBusiness.readProductsAsync(0, 20, queryString);
        }
        var createBusiness = function (app, businessName, onDone) {
            handle = app.isCreateBusinessPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        app.isCreateBusinessPending.unsubscribe(handle);
                        console.log("Creating Businesss complete");
                        onDone();
                    }
                }
            }));
            console.log("Creating Business");
            app.createBusinessAsync(businessName);
        }
        var readBusinesss = function (app, queryString, onDone) {
            handle = app.isBusinessesPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        app.isBusinessesPending.unsubscribe(handle);
                        console.log("Reading businesses complete");
                        onDone();
                    }
                }
            }));
            console.log("Reading businesses");
            app.readBusinessesAsync(0, 20, queryString);
        }

        var app = new module.AppFactory.getInstance().create();
        readBusinesss(app, "", function () {
            assert.equal(app.businesses.size(), 0, "Expected 0 businesses at first");
            createBusiness(app, "Business", function () {
                var business = app.createdBusiness.get();
                readBusinesss(app, "", function () {
                    assert.equal(app.businesses.size(), 1, "Expected 1 businesses after create");
                    readProducts(business, "", function () {
                        assert.equal(business.products.size(), 0, "Expected 1 products after create");
                        createProduct(business, "Product 1", function () {
                            readProducts(business, "", function () {
                                assert.equal(business.products.size(), 1, "Expected 1 products after create");
                                createProduct(business, "Product 2", function () {
                                    readProducts(business, "", function () {
                                        assert.equal(business.products.size(), 2, "Expected 1 products after create");
                                        done();
                                    });
                                });
                            });
                        });
                    });
                });
            });
        });
    });

    it('Create product on business out of scope', function (done) {
        var handle = null;

        var createProduct = function (newBusiness, productName, onDone) {
            handle = newBusiness.isCreateProductPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        newBusiness.isCreateProductPending.unsubscribe(handle);
                        console.log("Create product complete");
                        onDone();
                    }
                }
            }));
            console.log("Creating product");
            newBusiness.createProductAsync(productName);
        }
        var createBusiness = function (app, businessName, onDone) {
            handle = app.isCreateBusinessPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        app.isCreateBusinessPending.unsubscribe(handle);
                        console.log("Creating Businesss complete");
                        onDone();
                    }
                }
            }));
            console.log("Creating Business");
            app.createBusinessAsync(businessName);
        }

        var app = new module.AppFactory.getInstance().create();
        var handle = app.createdBusiness.subscribe(new module.ISubscribeHandleBusinessCallback.implement({
            onChange(newBusiness) {
                if (newBusiness != null) {
                    app.createdBusiness.unsubscribe(handle);
                    setTimeout(function () {
                        assert.throws(function () {
                            newBusiness.createProductAsync("Product");
                        }, "Expected create product to throw due to newBusiness being out of scope");
                        done();
                    }, 100);
                }
            }
        }));
        createBusiness(app, "Business", function () {        });
    });

    it('Clone subscription', function (done) {
        var handle = null;

        var createProduct = function (newBusiness, productName, onDone) {
            handle = newBusiness.isCreateProductPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        newBusiness.isCreateProductPending.unsubscribe(handle);
                        console.log("Reading products complete");
                        onDone();
                    }
                }
            }));
            console.log("Creating product");
            newBusiness.createProductAsync(productName);
        }
        var createBusiness = function (app, businessName, onDone) {
            handle = app.isCreateBusinessPending.subscribe(new module.ISubscribeHandleBoolCallback.implement({
                onChange(isPending) {
                    if (!isPending) {
                        app.isCreateBusinessPending.unsubscribe(handle);
                        console.log("Creating Businesss complete");
                        onDone();
                    }
                }
            }));
            console.log("Creating Business");
            app.createBusinessAsync(businessName);
        }

        var app = new module.AppFactory.getInstance().create();
        app.createdBusiness.subscribe(new module.ISubscribeHandleBusinessCallback.implement({
            onChange(newBusiness) {
                if (newBusiness != null) {
                    newBusiness = newBusiness.clone();
                    setTimeout(function () {
                        createProduct(newBusiness, "Product", function () {
                            newBusiness.delete();
                            assert.throws(function () {
                                newBusiness.createProductAsync("Product after deleted");
                            }, "Expected create product to throw due to newBusiness being out of scope");
                            done();
                        });
                    }, 100);
                }
            }
        }));
        createBusiness(app, "Business", function () {
        });
    });
});

File: Threadily.test.windows/AppExampleUnitTest.cpp
#include "stdafx.h"
#include "CppUnitTest.h"

#include "ThreadIds.h"
#include "ReadyEvent.h"

#include "AppFactory.h"
#include "App.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace threadily {
	namespace test {
		TEST_CLASS(AppExampleUnitTest)
		{
		public:

			/**
			* Basically tries to create everything in the system and any linkages in the system
			*/
			TEST_METHOD(ReadBusinessesAsync_VerifyNewBusinessIsOnUiThread)
			{
				auto app = AppFactory::getInstance().create();

				// first see if the business exists
				waitForAsync(
					app->isBusinessesPending,
					[&app]() {
					app->readBusinessesAsync(0, 10, "Wit");
				});
				Assert::AreEqual(0, (int)app->businesses->size(), L"Should be 0 businesses in the system");

				// Since it doesn't exist, we should make it!
				waitForAsync(
					app->isCreateBusinessPending,
					[&app]() {
					app->createBusinessAsync("Witness");
				});
				auto witness = app->createdBusiness->get();
				Assert::AreEqual(std::string("Witness"), witness->name->get());

				auto similarToThreadilyBridge = witness->products->subscribe([](std::shared_ptr<Product> newValue, size_t index, threadily::ObservableActionType action) {
				});

				// Now let's re-query that business again and check that it's on the right thread
				waitForAsync(
					app->isBusinessesPending,
					[&app]() {
					app->readBusinessesAsync(0, 10, "Wit");
				});
				Assert::AreEqual(1, (int)app->businesses->size(), L"Should be 1 businesses in the system");
				Assert::AreEqual((int)ThreadIds::ThreadId_UI, (int)app->businesses->at(0)->getThreadId(), L"Verify the business is on the UI thread");
			}

			/**
			* Basically tries to create everything in the system and any linkages in the system
			*/
			TEST_METHOD(ReadProductsAsync_bug_2017_03_18)
			{
				auto app = AppFactory::getInstance().create();

				// first see if the business exists
				waitForAsync(
					app->isBusinessesPending,
					[&app]() {
					app->readBusinessesAsync(0, 10, "Wit");
				});
				Assert::AreEqual(0, (int)app->businesses->size(), L"Should be 0 businesses in the system");

				// Since it doesn't exist, we should make it!
				waitForAsync(
					app->isCreateBusinessPending,
					[&app]() {
					app->createBusinessAsync("Witness");
				});
				auto witness = app->createdBusiness->get();
				Assert::AreEqual(std::string("Witness"), witness->name->get());

				auto similarToThreadilyBridge = witness->products->subscribe([](std::shared_ptr<Product> newValue, size_t index, threadily::ObservableActionType action) {
				});

				// Even though we just created the business let's check it's product listings anyway
				waitForAsync(
					witness->isProductsPending,
					[&witness]() {
					witness->readProductsAsync(0, 20, "");
				});
				Assert::AreEqual(0, (int)witness->products->size(), L"Should be 0 products for the business in the system");

				// create a product
				waitForAsync(
					witness->isCreateProductPending,
					[&witness]() {
					witness->createProductAsync("Name");
				});
				Assert::IsTrue(witness->createdProduct->get() != nullptr, L"Product should not be null after creating it");

				// Make sure the product exists
				waitForAsync(
					witness->isProductsPending,
					[&witness]() {
					witness->readProductsAsync(0, 20, "");
				});
				Assert::AreEqual(1, (int)witness->products->size(), L"Should be 1 products for the business in the system");
			}

			// https://github.com/k2snowman69/Threadily/issues/7
			TEST_METHOD(BusinessVectorIncreasesRefCount_issue_7)
			{
				auto app = AppFactory::getInstance().create();

				// Create a business
				waitForAsync(
					app->isCreateBusinessPending,
					[&app]() {
					app->createBusinessAsync("Witness");
				});
				auto witness = app->createdBusiness->get();
				Assert::AreEqual(std::string("Witness"), witness->name->get());

				auto oldUseCount = witness.use_count();

				// Now query the business populating app->businesses
				waitForAsync(
					app->isBusinessesPending,
					[&app]() {
					app->readBusinessesAsync(0, 10, "Wit");
				});
				Assert::AreEqual(1, (int)app->businesses->size(), L"Should be 1 businesses in the system");

				Assert::AreEqual(oldUseCount + 1, witness.use_count(), L"Expected the use count to increase by one as it should now be stored in the businesses listing");
			}

			void waitForAsync(std::shared_ptr<threadily::Observable<bool>> isPending, std::function<void()> asyncMethod)
			{
				threadily::ReadyEvent ready;
				auto subscriptionHandle = isPending->subscribe([&ready](bool isPending) {
					if (!isPending)
					{
						ready.finished();
					}
				});
				asyncMethod();
				ready.wait();
				isPending->unsubscribe(subscriptionHandle);
			}
		};
	}
}